litr/src/main/java/com/linkedin/android/litr/io/AudioRecordMediaSource.kt

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2022 LinkedIn Corporation
+ * All Rights Reserved.
+ *
+ * Licensed under the BSD 2-Clause License (the "License").  See License in the project root for
+ * license information.
+ *
+ * Author: Ian Bird
+ */
+package com.linkedin.android.litr.io
+
+import android.media.*
+import android.os.Build
+import android.util.Log
+import androidx.annotation.RequiresApi
+import com.linkedin.android.litr.MimeType
+import com.linkedin.android.litr.exception.MediaSourceException
+import java.nio.ByteBuffer
+
+private const val TAG = "AudioRecordMediaSource"
+
+private const val MEDIA_FORMAT_PCM_KEY = "pcm-encoding"
+private const val AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT
+private const val BYTES_PER_SAMPLE = 2
+
+/**
+ * An implementation of MediaSource, which utilizes Android's {@link AudioRecord} to record audio
+ * from a given audio source (i.e. the device's microphone).
+ *
+ * This class requires Android M+ in order to support reading data from {@link AudioRecord} in a
+ * non-blocking way.
+ */
+@RequiresApi(Build.VERSION_CODES.M)
+class AudioRecordMediaSource(
+        private val audioSource: Int = DEFAULT_AUDIO_SOURCE,
+        private val sampleRate: Int = DEFAULT_SAMPLE_RATE,
+        private val channelCount: Int = DEFAULT_CHANNEL_COUNT
+) : MediaSource {
+
+    init {
+        // Check that we've been configured with a supported number of channels.
+        if (channelCount !in 1..2) {
+            throw IllegalArgumentException("Unsupported channel count configuration")
+        }
+    }
+
+    private val channelConfig =
```

Comment:
we could probably do something like this here:
private val channelConfig = when (channelCount) {
   1 -> AudioFormat.CHANNEL_IN_MONO
   2 -> AudioFormat.CHANNEL_IN_STEREO
   else -> throw IllegalArgumentException("Unsupported channel count configuration")
}

This would get rid of init and make `AudioFormat` selection more explicit and readable.